### PR TITLE
Support MRP arrays with small amounts of invalid submaps

### PR DESCRIPTION
**Pull request by AndreiBarsan**
Invalid submaps are the two submaps for which we do not have submap UTMs. They affect very little data but we want to still index the underlying logs, and just leave gaps in UTM-ified MRP wherever the submap UTM is unavailable.

Also adds altitude to UTM methods for convenience (just the MRP altitude passed through). Worth noting that the MRP altitude roughly but not exactly corresponds to altitude above sea level.

Solves #44.

### CODE REVIEW

File: pit30m/monkeywrench.py [hunk 1]
```diff
@@ -237,13 +237,12 @@ def index(self, log_id: str, out_index_fpath: Optional[str] = None, check: bool
...
                                 check its shape too, etc.
         """
-        # XXX(andrei): This function is deprecated.
         self.index_all_cameras(log_id=log_id, out_index_fpath=out_index_fpath, check=check)
         self.index_lidar(log_id=log_id, out_index_fpath=out_index_fpath, check=check)
         res = self.diagnose_misc(log_id=log_id, out_index_fpath=out_index_fpath, check=check)
-        # XXX(andrei): Turn the misc diagnosis into a mini report too.
-        if len(res) > 0:
-            print("WARNING: misc non-sensor data had errors. See above for more information.")
+        # # TODO(andrei): Turn the misc diagnosis into a mini report too.
```
Comment (AndreiBarsan) [inline pit30m/monkeywrench.py:243]: This was unused before BTW

**Comment by AndreiBarsan** [comment]
CI will fail until we add the (sanitized) poses back in the bucket.

**Comment by AndreiBarsan** [comment]
@una-dinosauria Now that we have proper poses in the bucket, this PR is ready to review!

**Comment by AndreiBarsan** [comment]
(assuming 327... ever would touch test-query - I have not checked)

File: README.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 [![Python CI Status](https://github.com/pit30m/pit30m/actions/workflows/ci.yaml/badge.svg)](https://github.com/pit30m/pit30m/actions/workflows/ci.yaml)
 [![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](./LICENSE)
-![PyPI](https://img.shields.io/pypi/v/pit30m)
+[![PyPI](https://img.shields.io/pypi/v/pit30m)](https://pypi.org/project/pit30m/)
```
Comment (una-dinosauria) [inline README.md:5]: Oh neat, we have a pypi entry!

File: pit30m/data/log_reader.py [hunk 3]
```diff
@@ -225,16 +273,14 @@ def stereo_calib(self):
...
 
         In practice, users should use the camera/LiDAR iterators instead.
-
-        TODO(andrei): Document dtype (users now have to manually check dtype to learn).
         """
         pose_fpath = os.path.join(self._log_root_uri, self._pose_fname)
-        with self.fs.open(pose_fpath, "rb") as in_compressed_f:
-            return np.load(in_compressed_f)["data"]
+        with self.fs.open(pose_fpath, "rb") as wgs84_f:
+            return np.load(wgs84_f)["data"]
```
Comment (una-dinosauria) [inline pit30m/data/log_reader.py:283]: This is not wgs data, right?

File: pit30m/data/log_reader.py [hunk 4]
```diff
@@ -225,16 +273,14 @@ def stereo_calib(self):
             return data
 
     @cached_property
-    def raw_pose_data(self) -> np.ndarray:
-        """Returns the raw pose array, which needs manual association with other data types. 100Hz.
+    def raw_pose_data(self) -> npt.NDArray[RAW_POSE_DTYPE]:
```
Comment (una-dinosauria) [inline pit30m/data/log_reader.py:276]: omg I didn't know you could do that! so cool! 😁

File: pit30m/data/log_reader.py [hunk 5]
```diff
@@ -52,6 +53,53 @@ class LiDARFrame:
 
 PARTITIONS_BASEPATH = "s3://pit30m/partitions/"
 
+# Original dtype for unified raw pose arrays. Regular users should be using specialized getters, such as those for
+# continuous or map-relative poses, not the raw poses. '?' entries are bool but this does not register for some arcane
+# reason.
```
Comment (una-dinosauria) [inline pit30m/data/log_reader.py:58]: What do you mean by "this does not register"?

File: pit30m/data/submap.py [hunk 6]
```diff
@@ -20,6 +19,28 @@ def __init__(self, submap_id):
...
+    # Found on 2023-02-20 -- 1a8132d0-2634-4ad0-e433-ce5987ef0120 submap UUID was missing
+    "c9e9e7a7-f1cb-4af8-c5c9-3a610cbcc20e": [
+        UUID("1a8132d0-2634-4ad0-e433-ce5987ef0120"),
+    ],
+}
+
+BLANK_UUID = UUID("00000000-0000-0000-0000-000000000000")
+
+# test-query poses have been sanitized from the dataset and are not available. The sanitization includes the submap ID,
+# which for these entries has been replaced wiht th a null UUID.
```
Comment (una-dinosauria) [inline pit30m/data/submap.py:38]:
```suggestion
# which for these entries has been replaced with a null UUID.
```

File: pit30m/data/log_reader.py [hunk 7]
```diff
@@ -318,24 +364,23 @@ def map_relative_poses_dense(self) -> np.ndarray:
...
-              - An n-by-2 array with the UTM xy coordinates of the poses
+                - An n-long boolean array indicating whether the poses are valid
+                - An n-by-3 array with the UTM xy coordinates and altitudes of the poses
         """
         mrp = self.map_relative_poses_dense
         xyzs = rfn.structured_to_unstructured(mrp[["x", "y", "z"]])
+
         # Handle submap IDs which were truncated upon encoded due to ending with a zero.
-        submaps = [UUID(bytes=submap_uuid_bytes.ljust(16, b"\x00")) for submap_uuid_bytes in mrp["submap_id"]]
+        submaps = [UUID(bytes=submap_uuid_bytes).ljust(16, b"\x00") for submap_uuid_bytes in mrp["submap_id"]]
```
Comment (una-dinosauria) [inline pit30m/data/log_reader.py:376]: This looks like a typo. I think the previous code was correct.
```suggestion
        submaps = [UUID(bytes=submap_uuid_bytes.ljust(16, b"\x00")) for submap_uuid_bytes in mrp["submap_id"]]
```

I think adding a test (maybe in the integration tests of log reader) that calls `utm_poses_dense` from a log with nuked poses would flag this -- I already wrote this for another PR, happy to add it separately myself.

File: pit30m/data/submap.py [hunk 8]
```diff
@@ -20,6 +19,28 @@ def __init__(self, submap_id):
...
+KNOWN_MISSING_SUBMAPS = [submap_id for submap_ids in LOG_ID_TO_MISSING_SUBMAPS.values() for submap_id in submap_ids] + [
+    BLANK_UUID
+]
```
Comment (una-dinosauria) [inline pit30m/data/submap.py:41]: I find this a bit hard to read. Consider using `list(itertools.chain(*list_of_lists))`, or converting the values of the `LOG_ID_TO_MISSING_SUBMAPS` dict to sets

File: tests/submap_test.py [hunk 9]
```diff
@@ -50,12 +56,74 @@ def test_wgs84():
...
+
+def test_anonymized_entry():
+    """Ensure that handling an entry with a null submap ID yields empty UTMs."""
+    blank_uuid = UUID("00000000-0000-0000-0000-000000000000")
+    map_ = Map()
+    mrp_xyz_zeros = np.array([
+        [0.0, 0.0, 0.0]
+    ])
+    utm_poses = map_.to_utm(mrp_xyz_zeros, [blank_uuid], strict=False)
+    assert np.allclose(utm_poses, np.array([[np.nan, np.nan, np.nan]]), equal_nan=True)
```
Comment (una-dinosauria) [inline tests/submap_test.py:78]: nit: maybe `np.array_equal` would be better in these tests than `np.allclose`

**Review by una-dinosauria — APPROVED** [review]
A few minor comments, but overall looks fantastic. Thanks for this!

File: pit30m/data/log_reader.py (hunk 4 above)
Comment (AndreiBarsan) [reply pit30m/data/log_reader.py:276]: Yep, just gotta actually run `mypy` now :P

File: pit30m/data/log_reader.py (hunk 5 above)
Comment (AndreiBarsan) [reply pit30m/data/log_reader.py:58]: Oh as in numpy doesn't "see" those fields as bool when you load the npy file and print the dtype. I clarified the comment.

File: pit30m/data/log_reader.py (hunk 7 above)
Comment (AndreiBarsan) [reply pit30m/data/log_reader.py]: Reverted & test added! Nice catch!

File: pit30m/data/submap.py (hunk 8 above)
Comment (AndreiBarsan) [reply pit30m/data/submap.py:41]: cleaned up with more comments!

File: tests/submap_test.py (hunk 9 above)
Comment (AndreiBarsan) [reply tests/submap_test.py:78]: Done

File: README.md (hunk 2 above)
Comment (AndreiBarsan) [reply README.md:5]: :tada:

**Comment by AndreiBarsan** [comment]
Addressed comments, let me know what you think. Will land soon otherwise!

File: pit30m/data/log_reader.py [hunk 10]
```diff
@@ -52,6 +53,53 @@ class LiDARFrame:
 
 PARTITIONS_BASEPATH = "s3://pit30m/partitions/"
 
+# Original dtype for unified raw pose arrays. Regular users should be using specialized getters, such as those for
+# continuous or map-relative poses, not the raw poses. '?' entries are bool but numpy represents them at '?' for an
```
Comment (una-dinosauria) [inline pit30m/data/log_reader.py:57]: Gotcha. Yeah, the "?" can be confusing, although it is well-documented: https://numpy.org/doc/stable/reference/arrays.dtypes.html#specifying-and-constructing-data-types
```suggestion
# continuous or map-relative poses, not the raw poses. '?' entries are bool but numpy represents them as '?' for an
```

**Review by una-dinosauria — APPROVED** [review]
Looks great!

**Comment by una-dinosauria** [comment]
All looks good. Please go ahead and merge at your convenience.

**Comment by una-dinosauria** [comment]
I clicked the wrong button, please ignore me closing the PR 😐

**Comment by AndreiBarsan** [comment]
Updated the comments to remind the reader that '?' means bool, lol

**Pull request merged**